sklearn/model_selection/_search_successive_halving.py

```diff
@@ -0,0 +1,863 @@
+        # or smaller than n_required_iterations.
+        n_possible_iterations = 1 + floor(log(self.max_budget_ // self.r_min_,
+                                              self.ratio))
+
+        if self.aggressive_elimination:
+            n_iterations = n_required_iterations
+        else:
+            n_iterations = min(n_possible_iterations, n_required_iterations)
+
+        if self.verbose:
+            print('n_iterations: {}'.format(n_iterations))
+            print('n_required_iterations: {}'.format(n_required_iterations))
+            print('n_possible_iterations: {}'.format(n_possible_iterations))
+            print('r_min_: {}'.format(self.r_min_))
+            print('max_budget_: {}'.format(self.max_budget_))
+            print('aggressive_elimination: {}'.format(
+                self.aggressive_elimination))
+            print('force_exhaust_budget: {}'.format(self.force_exhaust_budget))
+            print('ratio: {}'.format(self.ratio))
+
+        self._r_i_list = []  # list of r_i for each iteration, used in tests
+
+        for iter_i in range(n_iterations):
+
+            power = iter_i  # default
+            if self.aggressive_elimination:
+                # this will set r_i to the initial value (i.e. the value of r_i
+                # at the first iteration) for as many iterations as needed
+                # (while candidates are being eliminated), and then go on as
+                # usual.
+                power = max(
+                    0,
+                    iter_i - n_required_iterations + n_possible_iterations
+                )
+
+            r_i = int(self.ratio**power * self.r_min_)
+            r_i = min(r_i, self.max_budget_)  # guard, probably not needed
+            self._r_i_list.append(r_i)
+
+            n_candidates = len(candidate_params)
+
+            if self.verbose:
+                print('-' * 10)
+                print('iter_i: {}'.format(iter_i))
+                print('n_candidates: {}'.format(n_candidates))
+                print('r_i: {}'.format(r_i))
+
+            if self.budget_on == 'n_samples':
+                stratify = y if is_classifier(self.estimator) else None
+                X_iter, y_iter = resample(X, y, replace=False,
+                                          random_state=rng, stratify=stratify,
+                                          n_samples=r_i)
+            else:
+                # Need copy so that r_i of next iteration do not overwrite
+                candidate_params = [c.copy() for c in candidate_params]
+                for candidate in candidate_params:
+                    candidate[self.budget_on] = r_i
+                X_iter, y_iter = X, y
+            more_results = {'iter': [iter_i] * n_candidates,
+                            'r_i': [r_i] * n_candidates}
+            results = evaluate_candidates(candidate_params, X_iter, y_iter,
+                                          more_results=more_results)
```

Comment:
Do we need this `more_results` mechanism? Can we not just update `results` directly in `_run_search`? or even support `_run_search` returning the final results?? How does `more_results` benefit us?